### ISSUE TITLE
Multipart.

### ISSUE THREAD

**Issue by rossabaker** [http4s/http4s#179]
- [x] Basic encoder
- [ ] Encoder that can calculate a content-length.
- [ ] Encoder that renders boundary header without extra ceremony.
- [x] Streaming decoder
- [ ] Handling of large parts.  Current working theory is a temp file.
- [x] multipart/\* model
- [ ] multipart/form model: convenient way to get parts by name, treat non-file items as a `UrlForm`, etc.

**Comment by jonoabroad** [comment]
I'm having a look at this, evenings only. If others are already working on it. Let me know.

**Comment by bryce-anderson** [comment]
Sounds great! Keep us posted.

**Comment by jonoabroad** [comment]
Hi, my  wife has just giving birth and so this ticket is going to go a little slower than expected from now. Current changes are in my fork and I'm hoping to process them as time permits. 
They are not ready for integration and are missing form header decoding.
Also I am beginning to think the implementation maybe naive.  
I'd like Naive and working, over nothing. 
Anyway - Let me know what peoples thoughts are, otherwise I'll progress when I can.
Cheers
Jono

**Comment by bryce-anderson** [comment]
Congratulations on the new addition to the family! I'll try take a look at your work in your fork as time permits.

**Comment by jonoabroad** [comment]
I should have said, it is in a branch -- https://github.com/jonoabroad/http4s/tree/issue/179

**Comment by refried** [comment]
I started down this path this morning too.  I think, in my current case, I can get away with using the HTML5 File API and a `String` body.

**Comment by jonoabroad** [comment]
I have the following, in which we parse the request into RawParts and then they can be `materialised` if we are interested in them, unfortunately this means reading in the entire part. Which is not memory friendly.
``` scala
final case class RawPart(body:List[Char]) extends Part {
   
  def materialise:Option[Part] = for {
     part   <- AttoPartParser().
                        part.
                        parse( body.mkString ).
                        done.option   
    } yield part     
}

final case class FormData(disposition:ContentDisposition,
                          content: Entity,
                          contentType: Option[ContentType] = None) extends Part {
  
  def materialise = Some(this)  
}

final case class Multipart(val parts: Process[Task, Part],
                           val boundary:Boundary = Boundary.create) {
```

While this allows streamed parsing, we can't inspect the headers before deciding if we want the body or not.  I think what we want is to read the headers of each part and use them to decide if we are interested in the body or if we want to drop it.

I am unsure what this will look look like, in terms of changes to `Multipart` and `parts`, but will look at it once back.

Any thoughts?

**Comment by beefyhalo** [comment]
Under my understanding Parboiled parsers will not work with binary data. It will force an encoding on the input.

**Comment by dbousamra** [comment]
Any update on this? I keep running into the need for multi-part uploads, and keep having to switch to Scalatra :(

**Comment by ritschwumm** [comment]
this is actually quite complicated - headers inside multipart contents follow rules quite different from the ones that apply to normal http headers. have a look at commons-fileupload, it seems they are trying hard to get it right, and it's a mess.

**Comment by rossabaker** [comment]
@jonoabroad, I know you've been working on this as life permits... would you like to post the latest, and we can get it on a branch for people to try and/or finish?

**Comment by jonoabroad** [comment]
Howdy, Im deepest darkest NSW this weekend.  I had a stab at using Atto, however based on @rossabaker suggestion I'm carving up the byteVector - a lot less byteVector <-> String translation. Let me get this tidied up and committed either this weekend or next. 

Sorry for the delays.

P.S @ritschwumm if I never have to see another RFC on form data it will be too soon.

**Comment by djspiewak** [comment]
Courtesy of RichRelevance.  This work was done quite a while ago, but the contribution was delayed because… I apparently… forgot?  Dunno.  Anyway, here it is: https://gist.github.com/djspiewak/9f079488e66778df0978

This revolves around the `Writer1` signature.  Attributes must be capable of existing entirely in memory (or at least, each individual attributes block must be able to exist in memory).  If I remember correctly, the RFC does allow for unbounded data within attributes blocks, and obviously this would be a pretty good way for someone to DoS a server, but as a practical matter I'm pretty sure that *most* web frameworks with multipart support have this same vulnerability.

Anyway, this transducer processes the stream and produces either a `Map[String, String]`, representing the attributes, or a `ByteVector`, representing a frame delimited chunk of data.  Only the current chunk (or attribute set) is retained in memory at a time, so it is possible to use this to write streaming upload handlers capable of ingesting arbitrarily large amounts of data (I believe the largest upload we tested was in the hundreds of GBs, but don't quote me; I know for a *fact* that we did a 70 GB test).

The main limitation is that this does not support separate multipart sections.  Once we start receiving the body, we're just going to blindly chunk through until we see the trailing delimiter, at which point we give up ever seeing anything ever again.  As a practical matter, this means that it does not support multiple file uploads in the same POST.  We also do not support *nested* multi-part, which might be the dumbest feature of any RFC I've ever seen and should die in a fire.  The former limitation seems more problematic, but also easier to fix (just go back to the start state when you see the closing delimiter and emit more attributes).  The latter case would need to be fixed to be RFC-compliant, but by the same token, the use-cases are quite limited (I honestly don't believe browsers will ever produce a nested multipart POST, and it's quite hard to do with `curl` as well).

Sorry for sitting on this for so long.  I got the clearance to open source this over a year ago, and I never posted it here.

**Comment by rossabaker** [comment]
Thanks, @djspiewak!  I dumped this gist into topic/form-parser to be sure that it compiles.  With one tweak to an inferred type, it does.  I will defer thinking about how this fits together with entity codecs until a day that didn't begin with wandering Philadelphia at 4am to catch a train.

**Comment by rossabaker** [comment]
This is a huge job, and one that we understand a lot better now.  I've broken it into a checklist.  Suggested edits to that list welcome.

**Comment by ChristopherDavenport** [comment]
I think that this should be closed and consolidated as we transition to `0.17` I think our current model works fairly well, if not for the 7th item. We can then reopen with remaining work on `0.17`

**Comment by rossabaker** [comment]
Works for me.  Nobody has screamed out for the unchecked features after all this time.
